// aws-cdk-lib/pipelines/lib/blueprint/file-set.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.FileSet = void 0;
const jsiiDeprecationWarnings = require("../../../.warnings.jsii.js");
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
/**
 * A set of files traveling through the deployment pipeline.
 *
 * Individual steps in the pipeline produce or consume
 * `FileSet`s.
 *
 * @stability stable
 */
class FileSet {
    /**
     * @param id Human-readable descriptor for this file set (does not need to be unique).
     * @stability stable
     */
    constructor(
    /** Human-readable descriptor for this file set (does not need to be unique) */
    id, producer) {
        this.id = id;
        /**
         * The primary output of a file set producer.
         *
         * The primary output of a FileSet is itself.
         *
         * @stability stable
         */
        this.primaryOutput = this;
        jsiiDeprecationWarnings.aws_cdk_lib_pipelines_Step(producer);
        this._producer = producer;
    }
    /**
     * The Step that produces this FileSet.
     *
     * @stability stable
     */
    get producer() {
        if (!this._producer) {
            throw new Error(`FileSet '${this.id}' doesn\'t have a producer; call 'fileSet.producedBy()'`);
        }
        return this._producer;
    }
    /**
     * Mark the given Step as the producer for this FileSet.
     *
     * This method can only be called once.
     *
     * @stability stable
     */
    producedBy(producer) {
        jsiiDeprecationWarnings.aws_cdk_lib_pipelines_Step(producer);
        if (this._producer) {
            throw new Error(`FileSet '${this.id}' already has a producer (${this._producer}) while setting producer: ${producer}`);
        }
        this._producer = producer;
    }
    /**
     * Return a string representation of this FileSet.
     *
     * @stability stable
     */
    toString() {
        return `FileSet(${this.id})`;
    }
}
exports.FileSet = FileSet;
_a = JSII_RTTI_SYMBOL_1;
FileSet[_a] = { fqn: "aws-cdk-lib.pipelines.FileSet", version: "2.0.0" };
