// aws-cdk-lib/pipelines/lib/blueprint/step.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.Step = void 0;
const jsiiDeprecationWarnings = require("../../../.warnings.jsii.js");
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const core_1 = require("../../../core");
/**
 * A generic Step which can be added to a Pipeline.
 *
 * Steps can be used to add Sources, Build Actions and Validations
 * to your pipeline.
 *
 * This class is abstract. See specific subclasses of Step for
 * useful steps to add to your Pipeline
 *
 * @stability stable
 */
class Step {
    /**
     * @param id Identifier for this step.
     * @stability stable
     */
    constructor(
    /** Identifier for this step */
    id) {
        this.id = id;
        /**
         * The list of FileSets consumed by this Step.
         *
         * @stability stable
         */
        this.dependencyFileSets = [];
        /**
         * Whether or not this is a Source step.
         *
         * What it means to be a Source step depends on the engine.
         *
         * @stability stable
         */
        this.isSource = false;
        if (core_1.Token.isUnresolved(id)) {
            throw new Error(`Step id cannot be unresolved, got '${id}'`);
        }
    }
    /**
     * Return the steps this step depends on, based on the FileSets it requires.
     *
     * @stability stable
     */
    get dependencies() {
        return this.dependencyFileSets.map(f => f.producer);
    }
    /**
     * Return a string representation of this Step.
     *
     * @stability stable
     */
    toString() {
        return `${this.constructor.name}(${this.id})`;
    }
    /**
     * The primary FileSet produced by this Step.
     *
     * Not all steps produce an output FileSet--if they do
     * you can substitute the `Step` object for the `FileSet` object.
     *
     * @stability stable
     */
    get primaryOutput() {
        // Accessor so it can be mutable in children
        return this._primaryOutput;
    }
    /**
     * Add an additional FileSet to the set of file sets required by this step.
     *
     * This will lead to a dependency on the producer of that file set.
     *
     * @stability stable
     */
    addDependencyFileSet(fs) {
        jsiiDeprecationWarnings.aws_cdk_lib_pipelines_FileSet(fs);
        this.dependencyFileSets.push(fs);
    }
    /**
     * Configure the given FileSet as the primary output of this step.
     *
     * @stability stable
     */
    configurePrimaryOutput(fs) {
        jsiiDeprecationWarnings.aws_cdk_lib_pipelines_FileSet(fs);
        this._primaryOutput = fs;
    }
}
exports.Step = Step;
_a = JSII_RTTI_SYMBOL_1;
Step[_a] = { fqn: "aws-cdk-lib.pipelines.Step", version: "2.0.0" };
