// aws-cdk-lib/pipelines/lib/blueprint/wave.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.Wave = void 0;
const jsiiDeprecationWarnings = require("../../../.warnings.jsii.js");
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const stage_deployment_1 = require("./stage-deployment");
/**
 * Multiple stages that are deployed in parallel.
 *
 * @stability stable
 */
class Wave {
    /**
     * @param id Identifier for this Wave.
     * @stability stable
     */
    constructor(
    /** Identifier for this Wave */
    id, props = {}) {
        var _b, _c;
        this.id = id;
        /**
         * The stages that are deployed in this wave.
         *
         * @stability stable
         */
        this.stages = [];
        jsiiDeprecationWarnings.aws_cdk_lib_pipelines_WaveProps(props);
        this.pre = (_b = props.pre) !== null && _b !== void 0 ? _b : [];
        this.post = (_c = props.post) !== null && _c !== void 0 ? _c : [];
    }
    /**
     * Add a Stage to this wave.
     *
     * It will be deployed in parallel with all other stages in this
     * wave.
     *
     * @stability stable
     */
    addStage(stage, options = {}) {
        jsiiDeprecationWarnings.aws_cdk_lib_Stage(stage);
        jsiiDeprecationWarnings.aws_cdk_lib_pipelines_AddStageOpts(options);
        const ret = stage_deployment_1.StageDeployment.fromStage(stage, options);
        this.stages.push(ret);
        return ret;
    }
    /**
     * Add an additional step to run before any of the stages in this wave.
     *
     * @stability stable
     */
    addPre(...steps) {
        jsiiDeprecationWarnings.aws_cdk_lib_pipelines_Step(steps);
        this.pre.push(...steps);
    }
    /**
     * Add an additional step to run after all of the stages in this wave.
     *
     * @stability stable
     */
    addPost(...steps) {
        jsiiDeprecationWarnings.aws_cdk_lib_pipelines_Step(steps);
        this.post.push(...steps);
    }
}
exports.Wave = Wave;
_a = JSII_RTTI_SYMBOL_1;
Wave[_a] = { fqn: "aws-cdk-lib.pipelines.Wave", version: "2.0.0" };
