// aws-cdk-lib/aws-elasticloadbalancingv2-actions/lib/cognito-action.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.AuthenticateCognitoAction = void 0;
const jsiiDeprecationWarnings = require("../../.warnings.jsii.js");
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const elbv2 = require("../../aws-elasticloadbalancingv2");
/**
 * A Listener Action to authenticate with Cognito.
 *
 * @stability stable
 */
class AuthenticateCognitoAction extends elbv2.ListenerAction {
    /**
     * Authenticate using an identity provide (IdP) that is compliant with OpenID Connect (OIDC).
     *
     * @stability stable
     */
    constructor(options) {
        var _b;
        super({
            type: 'authenticate-cognito',
            authenticateCognitoConfig: {
                userPoolArn: options.userPool.userPoolArn,
                userPoolClientId: options.userPoolClient.userPoolClientId,
                userPoolDomain: options.userPoolDomain.domainName,
                authenticationRequestExtraParams: options.authenticationRequestExtraParams,
                onUnauthenticatedRequest: options.onUnauthenticatedRequest,
                scope: options.scope,
                sessionCookieName: options.sessionCookieName,
                sessionTimeout: (_b = options.sessionTimeout) === null || _b === void 0 ? void 0 : _b.toSeconds().toString(),
            },
        }, options.next);
        jsiiDeprecationWarnings.aws_cdk_lib_aws_elasticloadbalancingv2_actions_AuthenticateCognitoActionProps(options);
    }
}
exports.AuthenticateCognitoAction = AuthenticateCognitoAction;
_a = JSII_RTTI_SYMBOL_1;
AuthenticateCognitoAction[_a] = { fqn: "aws-cdk-lib.aws_elasticloadbalancingv2_actions.AuthenticateCognitoAction", version: "2.0.0" };
