// aws-cdk-lib/aws-elasticloadbalancingv2-targets/lib/alb-target.js
"use strict";
var _a, _b;
Object.defineProperty(exports, "__esModule", { value: true });
exports.AlbTarget = exports.AlbArnTarget = void 0;
const jsiiDeprecationWarnings = require("../../.warnings.jsii.js");
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const elbv2 = require("../../aws-elasticloadbalancingv2");
/**
 * A single Application Load Balancer as the target for load balancing.
 *
 * @stability stable
 */
class AlbArnTarget {
    /**
     * Create a new alb target.
     *
     * @param albArn The ARN of the application load balancer to load balance to.
     * @param port The port on which the target is listening.
     * @stability stable
     */
    constructor(albArn, port) {
        this.albArn = albArn;
        this.port = port;
    }
    /**
     * Register this alb target with a load balancer.
     *
     * Don't call this, it is called automatically when you add the target to a
     * load balancer.
     *
     * @stability stable
     */
    attachToNetworkTargetGroup(targetGroup) {
        jsiiDeprecationWarnings.aws_cdk_lib_aws_elasticloadbalancingv2_INetworkTargetGroup(targetGroup);
        return this.attach(targetGroup);
    }
    attach(_targetGroup) {
        return {
            targetType: elbv2.TargetType.ALB,
            targetJson: { id: this.albArn, port: this.port },
        };
    }
}
exports.AlbArnTarget = AlbArnTarget;
_a = JSII_RTTI_SYMBOL_1;
AlbArnTarget[_a] = { fqn: "aws-cdk-lib.aws_elasticloadbalancingv2_targets.AlbArnTarget", version: "2.0.0" };
/**
 * A single Application Load Balancer as the target for load balancing.
 *
 * @stability stable
 */
class AlbTarget extends AlbArnTarget {
    /**
     * @param alb The application load balancer to load balance to.
     * @param port The port on which the target is listening.
     * @stability stable
     */
    constructor(alb, port) {
        super(alb.loadBalancerArn, port);
        jsiiDeprecationWarnings.aws_cdk_lib_aws_elasticloadbalancingv2_ApplicationLoadBalancer(alb);
    }
}
exports.AlbTarget = AlbTarget;
_b = JSII_RTTI_SYMBOL_1;
AlbTarget[_b] = { fqn: "aws-cdk-lib.aws_elasticloadbalancingv2_targets.AlbTarget", version: "2.0.0" };
