// aws-cdk-lib/aws-events-targets/lib/codepipeline.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.CodePipeline = void 0;
const jsiiDeprecationWarnings = require("../../.warnings.jsii.js");
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const iam = require("../../aws-iam");
const util_1 = require("./util");
/**
 * Allows the pipeline to be used as an EventBridge rule target.
 *
 * @stability stable
 */
class CodePipeline {
    /**
     * @stability stable
     */
    constructor(pipeline, options = {}) {
        this.pipeline = pipeline;
        this.options = options;
        jsiiDeprecationWarnings.aws_cdk_lib_aws_codepipeline_IPipeline(pipeline);
        jsiiDeprecationWarnings.aws_cdk_lib_aws_events_targets_CodePipelineTargetOptions(options);
    }
    /**
     * Returns the rule target specification.
     *
     * NOTE: Do not use the various `inputXxx` options. They can be set in a call to `addTarget`.
     *
     * @stability stable
     */
    bind(_rule, _id) {
        jsiiDeprecationWarnings.aws_cdk_lib_aws_events_IRule(_rule);
        return {
            ...util_1.bindBaseTargetConfig(this.options),
            id: '',
            arn: this.pipeline.pipelineArn,
            role: this.options.eventRole || util_1.singletonEventRole(this.pipeline, [new iam.PolicyStatement({
                    resources: [this.pipeline.pipelineArn],
                    actions: ['codepipeline:StartPipelineExecution'],
                })]),
            targetResource: this.pipeline,
        };
    }
}
exports.CodePipeline = CodePipeline;
_a = JSII_RTTI_SYMBOL_1;
CodePipeline[_a] = { fqn: "aws-cdk-lib.aws_events_targets.CodePipeline", version: "2.0.0" };
