// aws-cdk-lib/aws-events-targets/lib/lambda.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.LambdaFunction = void 0;
const jsiiDeprecationWarnings = require("../../.warnings.jsii.js");
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const util_1 = require("./util");
/**
 * Use an AWS Lambda function as an event rule target.
 *
 * @stability stable
 */
class LambdaFunction {
    /**
     * @stability stable
     */
    constructor(handler, props = {}) {
        this.handler = handler;
        this.props = props;
        jsiiDeprecationWarnings.aws_cdk_lib_aws_lambda_IFunction(handler);
        jsiiDeprecationWarnings.aws_cdk_lib_aws_events_targets_LambdaFunctionProps(props);
    }
    /**
     * Returns a RuleTarget that can be used to trigger this Lambda as a result from an EventBridge event.
     *
     * @stability stable
     */
    bind(rule, _id) {
        jsiiDeprecationWarnings.aws_cdk_lib_aws_events_IRule(rule);
        // Allow handler to be called from rule
        util_1.addLambdaPermission(rule, this.handler);
        if (this.props.deadLetterQueue) {
            util_1.addToDeadLetterQueueResourcePolicy(rule, this.props.deadLetterQueue);
        }
        return {
            ...util_1.bindBaseTargetConfig(this.props),
            arn: this.handler.functionArn,
            input: this.props.event,
            targetResource: this.handler,
        };
    }
}
exports.LambdaFunction = LambdaFunction;
_a = JSII_RTTI_SYMBOL_1;
LambdaFunction[_a] = { fqn: "aws-cdk-lib.aws_events_targets.LambdaFunction", version: "2.0.0" };
