// aws-cdk-lib/aws-events-targets/lib/sns.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.SnsTopic = void 0;
const jsiiDeprecationWarnings = require("../../.warnings.jsii.js");
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const iam = require("../../aws-iam");
/**
 * Use an SNS topic as a target for Amazon EventBridge rules.
 *
 * @stability stable
 * @example
 *
 *   /// fixture=withRepoAndTopic
 *   // publish to an SNS topic every time code is committed
 *   // to a CodeCommit repository
 *   repository.onCommit('onCommit', { target: new targets.SnsTopic(topic) });
 */
class SnsTopic {
    /**
     * @stability stable
     */
    constructor(topic, props = {}) {
        this.topic = topic;
        this.props = props;
        jsiiDeprecationWarnings.aws_cdk_lib_aws_sns_ITopic(topic);
        jsiiDeprecationWarnings.aws_cdk_lib_aws_events_targets_SnsTopicProps(props);
    }
    /**
     * Returns a RuleTarget that can be used to trigger this SNS topic as a result from an EventBridge event.
     *
     * @see https://docs.aws.amazon.com/eventbridge/latest/userguide/resource-based-policies-eventbridge.html#sns-permissions
     * @stability stable
     */
    bind(_rule, _id) {
        jsiiDeprecationWarnings.aws_cdk_lib_aws_events_IRule(_rule);
        // deduplicated automatically
        this.topic.grantPublish(new iam.ServicePrincipal('events.amazonaws.com'));
        return {
            arn: this.topic.topicArn,
            input: this.props.message,
            targetResource: this.topic,
        };
    }
}
exports.SnsTopic = SnsTopic;
_a = JSII_RTTI_SYMBOL_1;
SnsTopic[_a] = { fqn: "aws-cdk-lib.aws_events_targets.SnsTopic", version: "2.0.0" };
