// aws-cdk-lib/aws-opensearchservice/lib/opensearch-access-policy.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.OpenSearchAccessPolicy = void 0;
const iam = require("../../aws-iam");
const cr = require("../../custom-resources");
/**
 * Creates LogGroup resource policies.
 */
class OpenSearchAccessPolicy extends cr.AwsCustomResource {
    constructor(scope, id, props) {
        const policyDocument = new iam.PolicyDocument({
            statements: props.accessPolicies,
        });
        super(scope, id, {
            resourceType: 'Custom::OpenSearchAccessPolicy',
            onUpdate: {
                action: 'updateDomainConfig',
                service: 'OpenSearch',
                parameters: {
                    DomainName: props.domainName,
                    AccessPolicies: JSON.stringify(policyDocument.toJSON()),
                },
                // this is needed to limit the response body, otherwise it exceeds the CFN 4k limit
                outputPaths: ['DomainConfig.AccessPolicies'],
                physicalResourceId: cr.PhysicalResourceId.of(`${props.domainName}AccessPolicy`),
            },
            policy: cr.AwsCustomResourcePolicy.fromStatements([new iam.PolicyStatement({ actions: ['es:UpdateDomainConfig'], resources: [props.domainArn] })]),
        });
    }
}
exports.OpenSearchAccessPolicy = OpenSearchAccessPolicy;
