// aws-cdk-lib/aws-ses-actions/lib/stop.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.Stop = void 0;
const jsiiDeprecationWarnings = require("../../.warnings.jsii.js");
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
/**
 * Terminates the evaluation of the receipt rule set and optionally publishes a notification to Amazon SNS.
 *
 * @stability stable
 */
class Stop {
    /**
     * @stability stable
     */
    constructor(props = {}) {
        this.props = props;
        jsiiDeprecationWarnings.aws_cdk_lib_aws_ses_actions_StopProps(props);
    }
    /**
     * Returns the receipt rule action specification.
     *
     * @stability stable
     */
    bind(_rule) {
        var _b;
        jsiiDeprecationWarnings.aws_cdk_lib_aws_ses_IReceiptRule(_rule);
        return {
            stopAction: {
                scope: 'RuleSet',
                topicArn: (_b = this.props.topic) === null || _b === void 0 ? void 0 : _b.topicArn,
            },
        };
    }
}
exports.Stop = Stop;
_a = JSII_RTTI_SYMBOL_1;
Stop[_a] = { fqn: "aws-cdk-lib.aws_ses_actions.Stop", version: "2.0.0" };
