// aws-cdk-lib/aws-lambda-nodejs/lib/types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Charset = exports.SourceMapMode = exports.LogLevel = void 0;
/**
 * Log level for esbuild.
 *
 * @stability stable
 */
var LogLevel;
(function (LogLevel) {
    LogLevel["INFO"] = "info";
    LogLevel["WARNING"] = "warning";
    LogLevel["ERROR"] = "error";
    LogLevel["SILENT"] = "silent";
})(LogLevel = exports.LogLevel || (exports.LogLevel = {}));
/**
 * SourceMap mode for esbuild.
 *
 * @see https://esbuild.github.io/api/#sourcemap
 * @stability stable
 */
var SourceMapMode;
(function (SourceMapMode) {
    SourceMapMode["DEFAULT"] = "default";
    SourceMapMode["EXTERNAL"] = "external";
    SourceMapMode["INLINE"] = "inline";
    SourceMapMode["BOTH"] = "both";
})(SourceMapMode = exports.SourceMapMode || (exports.SourceMapMode = {}));
/**
 * Charset for esbuild's output.
 *
 * @stability stable
 */
var Charset;
(function (Charset) {
    Charset["ASCII"] = "ascii";
    Charset["UTF8"] = "utf8";
})(Charset = exports.Charset || (exports.Charset = {}));
