// aws-cdk-lib/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !exports.hasOwnProperty(p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.alexa_ask = require("./alexa-ask");
exports.assertions = require("./assertions");
exports.assets = require("./assets");
exports.aws_accessanalyzer = require("./aws-accessanalyzer");
exports.aws_acmpca = require("./aws-acmpca");
exports.aws_amazonmq = require("./aws-amazonmq");
exports.aws_amplify = require("./aws-amplify");
exports.aws_apigateway = require("./aws-apigateway");
exports.aws_apigatewayv2 = require("./aws-apigatewayv2");
exports.aws_appconfig = require("./aws-appconfig");
exports.aws_appflow = require("./aws-appflow");
exports.aws_appintegrations = require("./aws-appintegrations");
exports.aws_applicationautoscaling = require("./aws-applicationautoscaling");
exports.aws_applicationinsights = require("./aws-applicationinsights");
exports.aws_appmesh = require("./aws-appmesh");
exports.aws_apprunner = require("./aws-apprunner");
exports.aws_appstream = require("./aws-appstream");
exports.aws_appsync = require("./aws-appsync");
exports.aws_aps = require("./aws-aps");
exports.aws_athena = require("./aws-athena");
exports.aws_auditmanager = require("./aws-auditmanager");
exports.aws_autoscaling = require("./aws-autoscaling");
exports.aws_autoscaling_common = require("./aws-autoscaling-common");
exports.aws_autoscaling_hooktargets = require("./aws-autoscaling-hooktargets");
exports.aws_autoscalingplans = require("./aws-autoscalingplans");
exports.aws_backup = require("./aws-backup");
exports.aws_batch = require("./aws-batch");
exports.aws_budgets = require("./aws-budgets");
exports.aws_cassandra = require("./aws-cassandra");
exports.aws_ce = require("./aws-ce");
exports.aws_certificatemanager = require("./aws-certificatemanager");
exports.aws_chatbot = require("./aws-chatbot");
exports.aws_cloud9 = require("./aws-cloud9");
exports.aws_cloudformation = require("./aws-cloudformation");
exports.aws_cloudfront = require("./aws-cloudfront");
exports.aws_cloudfront_origins = require("./aws-cloudfront-origins");
exports.aws_cloudtrail = require("./aws-cloudtrail");
exports.aws_cloudwatch = require("./aws-cloudwatch");
exports.aws_cloudwatch_actions = require("./aws-cloudwatch-actions");
exports.aws_codeartifact = require("./aws-codeartifact");
exports.aws_codebuild = require("./aws-codebuild");
exports.aws_codecommit = require("./aws-codecommit");
exports.aws_codedeploy = require("./aws-codedeploy");
exports.aws_codeguruprofiler = require("./aws-codeguruprofiler");
exports.aws_codegurureviewer = require("./aws-codegurureviewer");
exports.aws_codepipeline = require("./aws-codepipeline");
exports.aws_codepipeline_actions = require("./aws-codepipeline-actions");
exports.aws_codestar = require("./aws-codestar");
exports.aws_codestarconnections = require("./aws-codestarconnections");
exports.aws_codestarnotifications = require("./aws-codestarnotifications");
exports.aws_cognito = require("./aws-cognito");
exports.aws_config = require("./aws-config");
exports.aws_connect = require("./aws-connect");
exports.aws_cur = require("./aws-cur");
exports.aws_customerprofiles = require("./aws-customerprofiles");
exports.aws_databrew = require("./aws-databrew");
exports.aws_datapipeline = require("./aws-datapipeline");
exports.aws_datasync = require("./aws-datasync");
exports.aws_dax = require("./aws-dax");
exports.aws_detective = require("./aws-detective");
exports.aws_devopsguru = require("./aws-devopsguru");
exports.aws_directoryservice = require("./aws-directoryservice");
exports.aws_dlm = require("./aws-dlm");
exports.aws_dms = require("./aws-dms");
exports.aws_docdb = require("./aws-docdb");
exports.aws_dynamodb = require("./aws-dynamodb");
exports.aws_ec2 = require("./aws-ec2");
exports.aws_ecr = require("./aws-ecr");
exports.aws_ecr_assets = require("./aws-ecr-assets");
exports.aws_ecs = require("./aws-ecs");
exports.aws_ecs_patterns = require("./aws-ecs-patterns");
exports.aws_efs = require("./aws-efs");
exports.aws_eks = require("./aws-eks");
exports.aws_elasticache = require("./aws-elasticache");
exports.aws_elasticbeanstalk = require("./aws-elasticbeanstalk");
exports.aws_elasticloadbalancing = require("./aws-elasticloadbalancing");
exports.aws_elasticloadbalancingv2 = require("./aws-elasticloadbalancingv2");
exports.aws_elasticloadbalancingv2_actions = require("./aws-elasticloadbalancingv2-actions");
exports.aws_elasticloadbalancingv2_targets = require("./aws-elasticloadbalancingv2-targets");
exports.aws_elasticsearch = require("./aws-elasticsearch");
exports.aws_emr = require("./aws-emr");
exports.aws_emrcontainers = require("./aws-emrcontainers");
exports.aws_events = require("./aws-events");
exports.aws_events_targets = require("./aws-events-targets");
exports.aws_eventschemas = require("./aws-eventschemas");
exports.aws_finspace = require("./aws-finspace");
exports.aws_fis = require("./aws-fis");
exports.aws_fms = require("./aws-fms");
exports.aws_frauddetector = require("./aws-frauddetector");
exports.aws_fsx = require("./aws-fsx");
exports.aws_gamelift = require("./aws-gamelift");
exports.aws_globalaccelerator = require("./aws-globalaccelerator");
exports.aws_globalaccelerator_endpoints = require("./aws-globalaccelerator-endpoints");
exports.aws_glue = require("./aws-glue");
exports.aws_greengrass = require("./aws-greengrass");
exports.aws_greengrassv2 = require("./aws-greengrassv2");
exports.aws_groundstation = require("./aws-groundstation");
exports.aws_guardduty = require("./aws-guardduty");
exports.aws_healthlake = require("./aws-healthlake");
exports.aws_iam = require("./aws-iam");
exports.aws_imagebuilder = require("./aws-imagebuilder");
exports.aws_inspector = require("./aws-inspector");
exports.aws_iot = require("./aws-iot");
exports.aws_iot1click = require("./aws-iot1click");
exports.aws_iotanalytics = require("./aws-iotanalytics");
exports.aws_iotcoredeviceadvisor = require("./aws-iotcoredeviceadvisor");
exports.aws_iotevents = require("./aws-iotevents");
exports.aws_iotfleethub = require("./aws-iotfleethub");
exports.aws_iotsitewise = require("./aws-iotsitewise");
exports.aws_iotthingsgraph = require("./aws-iotthingsgraph");
exports.aws_iotwireless = require("./aws-iotwireless");
exports.aws_ivs = require("./aws-ivs");
exports.aws_kendra = require("./aws-kendra");
exports.aws_kinesis = require("./aws-kinesis");
exports.aws_kinesisanalytics = require("./aws-kinesisanalytics");
exports.aws_kinesisfirehose = require("./aws-kinesisfirehose");
exports.aws_kms = require("./aws-kms");
exports.aws_lakeformation = require("./aws-lakeformation");
exports.aws_lambda = require("./aws-lambda");
exports.aws_lambda_destinations = require("./aws-lambda-destinations");
exports.aws_lambda_event_sources = require("./aws-lambda-event-sources");
exports.aws_lambda_nodejs = require("./aws-lambda-nodejs");
exports.aws_licensemanager = require("./aws-licensemanager");
exports.aws_lightsail = require("./aws-lightsail");
exports.aws_location = require("./aws-location");
exports.aws_logs = require("./aws-logs");
exports.aws_logs_destinations = require("./aws-logs-destinations");
exports.aws_lookoutequipment = require("./aws-lookoutequipment");
exports.aws_lookoutmetrics = require("./aws-lookoutmetrics");
exports.aws_lookoutvision = require("./aws-lookoutvision");
exports.aws_macie = require("./aws-macie");
exports.aws_managedblockchain = require("./aws-managedblockchain");
exports.aws_mediaconnect = require("./aws-mediaconnect");
exports.aws_mediaconvert = require("./aws-mediaconvert");
exports.aws_medialive = require("./aws-medialive");
exports.aws_mediapackage = require("./aws-mediapackage");
exports.aws_mediastore = require("./aws-mediastore");
exports.aws_memorydb = require("./aws-memorydb");
exports.aws_msk = require("./aws-msk");
exports.aws_mwaa = require("./aws-mwaa");
exports.aws_neptune = require("./aws-neptune");
exports.aws_networkfirewall = require("./aws-networkfirewall");
exports.aws_networkmanager = require("./aws-networkmanager");
exports.aws_nimblestudio = require("./aws-nimblestudio");
exports.aws_opensearchservice = require("./aws-opensearchservice");
exports.aws_opsworks = require("./aws-opsworks");
exports.aws_opsworkscm = require("./aws-opsworkscm");
exports.aws_panorama = require("./aws-panorama");
exports.aws_pinpoint = require("./aws-pinpoint");
exports.aws_pinpointemail = require("./aws-pinpointemail");
exports.aws_qldb = require("./aws-qldb");
exports.aws_quicksight = require("./aws-quicksight");
exports.aws_ram = require("./aws-ram");
exports.aws_rds = require("./aws-rds");
exports.aws_redshift = require("./aws-redshift");
exports.aws_rekognition = require("./aws-rekognition");
exports.aws_resourcegroups = require("./aws-resourcegroups");
exports.aws_robomaker = require("./aws-robomaker");
exports.aws_route53 = require("./aws-route53");
exports.aws_route53_patterns = require("./aws-route53-patterns");
exports.aws_route53_targets = require("./aws-route53-targets");
exports.aws_route53recoverycontrol = require("./aws-route53recoverycontrol");
exports.aws_route53recoveryreadiness = require("./aws-route53recoveryreadiness");
exports.aws_route53resolver = require("./aws-route53resolver");
exports.aws_s3 = require("./aws-s3");
exports.aws_s3_assets = require("./aws-s3-assets");
exports.aws_s3_deployment = require("./aws-s3-deployment");
exports.aws_s3_notifications = require("./aws-s3-notifications");
exports.aws_s3objectlambda = require("./aws-s3objectlambda");
exports.aws_s3outposts = require("./aws-s3outposts");
exports.aws_sagemaker = require("./aws-sagemaker");
exports.aws_sam = require("./aws-sam");
exports.aws_sdb = require("./aws-sdb");
exports.aws_secretsmanager = require("./aws-secretsmanager");
exports.aws_securityhub = require("./aws-securityhub");
exports.aws_servicecatalog = require("./aws-servicecatalog");
exports.aws_servicecatalogappregistry = require("./aws-servicecatalogappregistry");
exports.aws_servicediscovery = require("./aws-servicediscovery");
exports.aws_ses = require("./aws-ses");
exports.aws_ses_actions = require("./aws-ses-actions");
exports.aws_signer = require("./aws-signer");
exports.aws_sns = require("./aws-sns");
exports.aws_sns_subscriptions = require("./aws-sns-subscriptions");
exports.aws_sqs = require("./aws-sqs");
exports.aws_ssm = require("./aws-ssm");
exports.aws_ssmcontacts = require("./aws-ssmcontacts");
exports.aws_ssmincidents = require("./aws-ssmincidents");
exports.aws_sso = require("./aws-sso");
exports.aws_stepfunctions = require("./aws-stepfunctions");
exports.aws_stepfunctions_tasks = require("./aws-stepfunctions-tasks");
exports.aws_synthetics = require("./aws-synthetics");
exports.aws_timestream = require("./aws-timestream");
exports.aws_transfer = require("./aws-transfer");
exports.aws_waf = require("./aws-waf");
exports.aws_wafregional = require("./aws-wafregional");
exports.aws_wafv2 = require("./aws-wafv2");
exports.aws_wisdom = require("./aws-wisdom");
exports.aws_workspaces = require("./aws-workspaces");
exports.aws_xray = require("./aws-xray");
exports.cloud_assembly_schema = require("./cloud-assembly-schema");
exports.cloudformation_include = require("./cloudformation-include");
__exportStar(require("./core"), exports);
exports.custom_resources = require("./custom-resources");
exports.cx_api = require("./cx-api");
exports.lambda_layer_awscli = require("./lambda-layer-awscli");
exports.lambda_layer_kubectl = require("./lambda-layer-kubectl");
exports.lambda_layer_node_proxy_agent = require("./lambda-layer-node-proxy-agent");
exports.pipelines = require("./pipelines");
exports.region_info = require("./region-info");
