// aws-cdk-lib/aws-cloudwatch-actions/lib/ec2.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.Ec2Action = exports.Ec2InstanceAction = void 0;
const jsiiDeprecationWarnings = require("../../.warnings.jsii.js");
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const core_1 = require("../../core");
/**
 * Types of EC2 actions available.
 *
 * @stability stable
 */
var Ec2InstanceAction;
(function (Ec2InstanceAction) {
    Ec2InstanceAction["STOP"] = "stop";
    Ec2InstanceAction["TERMINATE"] = "terminate";
    Ec2InstanceAction["RECOVER"] = "recover";
    Ec2InstanceAction["REBOOT"] = "reboot";
})(Ec2InstanceAction = exports.Ec2InstanceAction || (exports.Ec2InstanceAction = {}));
/**
 * Use an EC2 action as an Alarm action.
 *
 * @stability stable
 */
class Ec2Action {
    /**
     * @stability stable
     */
    constructor(instanceAction) {
        jsiiDeprecationWarnings.aws_cdk_lib_aws_cloudwatch_actions_Ec2InstanceAction(instanceAction);
        this.ec2Action = instanceAction;
    }
    /**
     * Returns an alarm action configuration to use an EC2 action as an alarm action.
     *
     * @stability stable
     */
    bind(_scope, _alarm) {
        jsiiDeprecationWarnings.aws_cdk_lib_aws_cloudwatch_IAlarm(_alarm);
        return { alarmActionArn: `arn:aws:automate:${core_1.Stack.of(_scope).region}:ec2:${this.ec2Action}` };
    }
}
exports.Ec2Action = Ec2Action;
_a = JSII_RTTI_SYMBOL_1;
Ec2Action[_a] = { fqn: "aws-cdk-lib.aws_cloudwatch_actions.Ec2Action", version: "2.0.0" };
