// aws-cdk-lib/aws-amazonmq/lib/amazonmq-canned-metrics.generated.js
"use strict";
// Copyright 2012-2021 Amazon.com, Inc. or its affiliates. All Rights Reserved.
Object.defineProperty(exports, "__esModule", { value: true });
exports.AmazonMQMetrics = void 0;
/* eslint-disable max-len */ // This is generated code - line lengths are difficult to control
class AmazonMQMetrics {
    static cpuUtilizationAverage(dimensions) {
        return {
            namespace: 'AWS/AmazonMQ',
            metricName: 'CpuUtilization',
            dimensionsMap: dimensions,
            statistic: 'Average',
        };
    }
    static currentConnectionsCountSum(dimensions) {
        return {
            namespace: 'AWS/AmazonMQ',
            metricName: 'CurrentConnectionsCount',
            dimensionsMap: dimensions,
            statistic: 'Sum',
        };
    }
    static cpuCreditBalanceHeapUsageMaximum(dimensions) {
        return {
            namespace: 'AWS/AmazonMQ',
            metricName: 'CpuCreditBalanceHeapUsage',
            dimensionsMap: dimensions,
            statistic: 'Maximum',
        };
    }
    static networkInSum(dimensions) {
        return {
            namespace: 'AWS/AmazonMQ',
            metricName: 'NetworkIn',
            dimensionsMap: dimensions,
            statistic: 'Sum',
        };
    }
    static networkOutSum(dimensions) {
        return {
            namespace: 'AWS/AmazonMQ',
            metricName: 'NetworkOut',
            dimensionsMap: dimensions,
            statistic: 'Sum',
        };
    }
    static totalConsumerCountSum(dimensions) {
        return {
            namespace: 'AWS/AmazonMQ',
            metricName: 'TotalConsumerCount',
            dimensionsMap: dimensions,
            statistic: 'Sum',
        };
    }
    static totalMessageCountSum(dimensions) {
        return {
            namespace: 'AWS/AmazonMQ',
            metricName: 'TotalMessageCount',
            dimensionsMap: dimensions,
            statistic: 'Sum',
        };
    }
    static totalProducerCountSum(dimensions) {
        return {
            namespace: 'AWS/AmazonMQ',
            metricName: 'TotalProducerCount',
            dimensionsMap: dimensions,
            statistic: 'Sum',
        };
    }
}
exports.AmazonMQMetrics = AmazonMQMetrics;
