// aws-cdk-lib/aws-sns-subscriptions/lib/sms.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.SmsSubscription = void 0;
const jsiiDeprecationWarnings = require("../../.warnings.jsii.js");
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const sns = require("../../aws-sns");
/**
 * Use an sms address as a subscription target.
 *
 * @stability stable
 */
class SmsSubscription {
    /**
     * @stability stable
     */
    constructor(phoneNumber, props = {}) {
        this.phoneNumber = phoneNumber;
        this.props = props;
        jsiiDeprecationWarnings.aws_cdk_lib_aws_sns_subscriptions_SmsSubscriptionProps(props);
    }
    /**
     * Returns a configuration used to subscribe to an SNS topic.
     *
     * @stability stable
     */
    bind(_topic) {
        jsiiDeprecationWarnings.aws_cdk_lib_aws_sns_ITopic(_topic);
        return {
            subscriberId: this.phoneNumber,
            endpoint: this.phoneNumber,
            protocol: sns.SubscriptionProtocol.SMS,
            filterPolicy: this.props.filterPolicy,
        };
    }
}
exports.SmsSubscription = SmsSubscription;
_a = JSII_RTTI_SYMBOL_1;
SmsSubscription[_a] = { fqn: "aws-cdk-lib.aws_sns_subscriptions.SmsSubscription", version: "2.0.0" };
