// aws-cdk-lib/assertions/lib/capture.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.Capture = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const matcher_1 = require("./matcher");
const type_1 = require("./private/type");
/**
 * Capture values while matching templates.
 *
 * Using an instance of this class within a Matcher will capture the matching value.
 * The `as*()` APIs on the instance can be used to get the captured value.
 *
 * @stability stable
 */
class Capture extends matcher_1.Matcher {
    /**
     * @stability stable
     */
    constructor() {
        super();
        this.value = null;
        this.name = 'Capture';
    }
    /**
     * Test whether a target matches the provided pattern.
     *
     * Every Matcher must implement this method.
     * This method will be invoked by the assertions framework. Do not call this method directly.
     *
     * @stability stable
     */
    test(actual) {
        this.value = actual;
        const result = new matcher_1.MatchResult(actual);
        if (actual == null) {
            result.push(this, [], `Can only capture non-nullish values. Found ${actual}`);
        }
        return result;
    }
    /**
     * Retrieve the captured value as a string.
     *
     * An error is generated if no value is captured or if the value is not a string.
     *
     * @stability stable
     */
    asString() {
        this.checkNotNull();
        if (type_1.getType(this.value) === 'string') {
            return this.value;
        }
        this.reportIncorrectType('string');
    }
    /**
     * Retrieve the captured value as a number.
     *
     * An error is generated if no value is captured or if the value is not a number.
     *
     * @stability stable
     */
    asNumber() {
        this.checkNotNull();
        if (type_1.getType(this.value) === 'number') {
            return this.value;
        }
        this.reportIncorrectType('number');
    }
    /**
     * Retrieve the captured value as a boolean.
     *
     * An error is generated if no value is captured or if the value is not a boolean.
     *
     * @stability stable
     */
    asBoolean() {
        this.checkNotNull();
        if (type_1.getType(this.value) === 'boolean') {
            return this.value;
        }
        this.reportIncorrectType('boolean');
    }
    /**
     * Retrieve the captured value as an array.
     *
     * An error is generated if no value is captured or if the value is not an array.
     *
     * @stability stable
     */
    asArray() {
        this.checkNotNull();
        if (type_1.getType(this.value) === 'array') {
            return this.value;
        }
        this.reportIncorrectType('array');
    }
    /**
     * Retrieve the captured value as a JSON object.
     *
     * An error is generated if no value is captured or if the value is not an object.
     *
     * @stability stable
     */
    asObject() {
        this.checkNotNull();
        if (type_1.getType(this.value) === 'object') {
            return this.value;
        }
        this.reportIncorrectType('object');
    }
    checkNotNull() {
        if (this.value == null) {
            throw new Error('No value captured');
        }
    }
    reportIncorrectType(expected) {
        throw new Error(`Captured value is expected to be ${expected} but found ${type_1.getType(this.value)}. ` +
            `Value is ${JSON.stringify(this.value, undefined, 2)}`);
    }
}
exports.Capture = Capture;
_a = JSII_RTTI_SYMBOL_1;
Capture[_a] = { fqn: "aws-cdk-lib.assertions.Capture", version: "2.0.0" };
