// aws-cdk-lib/aws-route53-targets/lib/api-gateway-domain-name.js
"use strict";
var _a, _b;
Object.defineProperty(exports, "__esModule", { value: true });
exports.ApiGateway = exports.ApiGatewayDomain = void 0;
const jsiiDeprecationWarnings = require("../../.warnings.jsii.js");
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
/**
 * Defines an API Gateway domain name as the alias target.
 *
 * Use the `ApiGateway` class if you wish to map the alias to an REST API with a
 * domain name defined through the `RestApiProps.domainName` prop.
 *
 * @stability stable
 */
class ApiGatewayDomain {
    /**
     * @stability stable
     */
    constructor(domainName) {
        this.domainName = domainName;
        jsiiDeprecationWarnings.aws_cdk_lib_aws_apigateway_IDomainName(domainName);
    }
    /**
     * Return hosted zone ID and DNS name, usable for Route53 alias targets.
     *
     * @stability stable
     */
    bind(_record, _zone) {
        jsiiDeprecationWarnings.aws_cdk_lib_aws_route53_IRecordSet(_record);
        jsiiDeprecationWarnings.aws_cdk_lib_aws_route53_IHostedZone(_zone);
        return {
            dnsName: this.domainName.domainNameAliasDomainName,
            hostedZoneId: this.domainName.domainNameAliasHostedZoneId,
        };
    }
}
exports.ApiGatewayDomain = ApiGatewayDomain;
_a = JSII_RTTI_SYMBOL_1;
ApiGatewayDomain[_a] = { fqn: "aws-cdk-lib.aws_route53_targets.ApiGatewayDomain", version: "2.0.0" };
/**
 * Defines an API Gateway REST API as the alias target. Requires that the domain name will be defined through `RestApiProps.domainName`.
 *
 * You can direct the alias to any `apigateway.DomainName` resource through the
 * `ApiGatewayDomain` class.
 *
 * @stability stable
 */
class ApiGateway extends ApiGatewayDomain {
    /**
     * @stability stable
     */
    constructor(api) {
        jsiiDeprecationWarnings.aws_cdk_lib_aws_apigateway_RestApiBase(api);
        if (!api.domainName) {
            throw new Error('API does not define a default domain name');
        }
        super(api.domainName);
    }
}
exports.ApiGateway = ApiGateway;
_b = JSII_RTTI_SYMBOL_1;
ApiGateway[_b] = { fqn: "aws-cdk-lib.aws_route53_targets.ApiGateway", version: "2.0.0" };
