// aws-cdk-lib/aws-dax/lib/dax-canned-metrics.generated.js
"use strict";
// Copyright 2012-2021 Amazon.com, Inc. or its affiliates. All Rights Reserved.
Object.defineProperty(exports, "__esModule", { value: true });
exports.DAXMetrics = void 0;
/* eslint-disable max-len */ // This is generated code - line lengths are difficult to control
class DAXMetrics {
    static cpuUtilizationAverage(dimensions) {
        return {
            namespace: 'AWS/DAX',
            metricName: 'CPUUtilization',
            dimensionsMap: dimensions,
            statistic: 'Average',
        };
    }
    static itemCacheHitsSum(dimensions) {
        return {
            namespace: 'AWS/DAX',
            metricName: 'ItemCacheHits',
            dimensionsMap: dimensions,
            statistic: 'Sum',
        };
    }
    static itemCacheMissesSum(dimensions) {
        return {
            namespace: 'AWS/DAX',
            metricName: 'ItemCacheMisses',
            dimensionsMap: dimensions,
            statistic: 'Sum',
        };
    }
    static batchGetItemRequestCountSum(dimensions) {
        return {
            namespace: 'AWS/DAX',
            metricName: 'BatchGetItemRequestCount',
            dimensionsMap: dimensions,
            statistic: 'Sum',
        };
    }
    static batchWriteItemRequestCountSum(dimensions) {
        return {
            namespace: 'AWS/DAX',
            metricName: 'BatchWriteItemRequestCount',
            dimensionsMap: dimensions,
            statistic: 'Sum',
        };
    }
    static cacheMemoryUtilizationAverage(dimensions) {
        return {
            namespace: 'AWS/DAX',
            metricName: 'CacheMemoryUtilization',
            dimensionsMap: dimensions,
            statistic: 'Average',
        };
    }
    static clientConnectionsSum(dimensions) {
        return {
            namespace: 'AWS/DAX',
            metricName: 'ClientConnections',
            dimensionsMap: dimensions,
            statistic: 'Sum',
        };
    }
    static deleteItemRequestCountSum(dimensions) {
        return {
            namespace: 'AWS/DAX',
            metricName: 'DeleteItemRequestCount',
            dimensionsMap: dimensions,
            statistic: 'Sum',
        };
    }
    static errorRequestCountSum(dimensions) {
        return {
            namespace: 'AWS/DAX',
            metricName: 'ErrorRequestCount',
            dimensionsMap: dimensions,
            statistic: 'Sum',
        };
    }
    static estimatedDbSizeAverage(dimensions) {
        return {
            namespace: 'AWS/DAX',
            metricName: 'EstimatedDbSize',
            dimensionsMap: dimensions,
            statistic: 'Average',
        };
    }
    static evictedSizeSum(dimensions) {
        return {
            namespace: 'AWS/DAX',
            metricName: 'EvictedSize',
            dimensionsMap: dimensions,
            statistic: 'Sum',
        };
    }
    static failedRequestCountSum(dimensions) {
        return {
            namespace: 'AWS/DAX',
            metricName: 'FailedRequestCount',
            dimensionsMap: dimensions,
            statistic: 'Sum',
        };
    }
    static faultRequestCountSum(dimensions) {
        return {
            namespace: 'AWS/DAX',
            metricName: 'FaultRequestCount',
            dimensionsMap: dimensions,
            statistic: 'Sum',
        };
    }
    static getItemRequestCountSum(dimensions) {
        return {
            namespace: 'AWS/DAX',
            metricName: 'GetItemRequestCount',
            dimensionsMap: dimensions,
            statistic: 'Sum',
        };
    }
    static networkBytesInSum(dimensions) {
        return {
            namespace: 'AWS/DAX',
            metricName: 'NetworkBytesIn',
            dimensionsMap: dimensions,
            statistic: 'Sum',
        };
    }
    static networkBytesOutSum(dimensions) {
        return {
            namespace: 'AWS/DAX',
            metricName: 'NetworkBytesOut',
            dimensionsMap: dimensions,
            statistic: 'Sum',
        };
    }
    static networkPacketsInSum(dimensions) {
        return {
            namespace: 'AWS/DAX',
            metricName: 'NetworkPacketsIn',
            dimensionsMap: dimensions,
            statistic: 'Sum',
        };
    }
    static networkPacketsOutSum(dimensions) {
        return {
            namespace: 'AWS/DAX',
            metricName: 'NetworkPacketsOut',
            dimensionsMap: dimensions,
            statistic: 'Sum',
        };
    }
    static putItemRequestCountSum(dimensions) {
        return {
            namespace: 'AWS/DAX',
            metricName: 'PutItemRequestCount',
            dimensionsMap: dimensions,
            statistic: 'Sum',
        };
    }
    static queryCacheHitsSum(dimensions) {
        return {
            namespace: 'AWS/DAX',
            metricName: 'QueryCacheHits',
            dimensionsMap: dimensions,
            statistic: 'Sum',
        };
    }
    static queryCacheMissesSum(dimensions) {
        return {
            namespace: 'AWS/DAX',
            metricName: 'QueryCacheMisses',
            dimensionsMap: dimensions,
            statistic: 'Sum',
        };
    }
    static queryRequestCountSum(dimensions) {
        return {
            namespace: 'AWS/DAX',
            metricName: 'QueryRequestCount',
            dimensionsMap: dimensions,
            statistic: 'Sum',
        };
    }
    static scanCacheHitsSum(dimensions) {
        return {
            namespace: 'AWS/DAX',
            metricName: 'ScanCacheHits',
            dimensionsMap: dimensions,
            statistic: 'Sum',
        };
    }
    static scanCacheMissesSum(dimensions) {
        return {
            namespace: 'AWS/DAX',
            metricName: 'ScanCacheMisses',
            dimensionsMap: dimensions,
            statistic: 'Sum',
        };
    }
    static scanRequestCountSum(dimensions) {
        return {
            namespace: 'AWS/DAX',
            metricName: 'ScanRequestCount',
            dimensionsMap: dimensions,
            statistic: 'Sum',
        };
    }
    static throttledRequestCountSum(dimensions) {
        return {
            namespace: 'AWS/DAX',
            metricName: 'ThrottledRequestCount',
            dimensionsMap: dimensions,
            statistic: 'Sum',
        };
    }
    static totalRequestCountSum(dimensions) {
        return {
            namespace: 'AWS/DAX',
            metricName: 'TotalRequestCount',
            dimensionsMap: dimensions,
            statistic: 'Sum',
        };
    }
    static transactGetItemsCountSum(dimensions) {
        return {
            namespace: 'AWS/DAX',
            metricName: 'TransactGetItemsCount',
            dimensionsMap: dimensions,
            statistic: 'Sum',
        };
    }
    static transactWriteItemsCountSum(dimensions) {
        return {
            namespace: 'AWS/DAX',
            metricName: 'TransactWriteItemsCount',
            dimensionsMap: dimensions,
            statistic: 'Sum',
        };
    }
    static updateItemRequestCountSum(dimensions) {
        return {
            namespace: 'AWS/DAX',
            metricName: 'UpdateItemRequestCount',
            dimensionsMap: dimensions,
            statistic: 'Sum',
        };
    }
}
exports.DAXMetrics = DAXMetrics;
