// aws-cdk-lib/aws-backup/lib/resource.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.BackupResource = exports.TagOperation = void 0;
const jsiiDeprecationWarnings = require("../../.warnings.jsii.js");
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const core_1 = require("../../core");
/**
 * An operation that is applied to a key-value pair.
 *
 * @stability stable
 */
var TagOperation;
(function (TagOperation) {
    TagOperation["STRING_EQUALS"] = "STRINGEQUALS";
    TagOperation["DUMMY"] = "dummy";
})(TagOperation = exports.TagOperation || (exports.TagOperation = {}));
/**
 * A resource to backup.
 *
 * @stability stable
 */
class BackupResource {
    /**
     * @stability stable
     */
    constructor(resource, tagCondition, construct) {
        jsiiDeprecationWarnings.aws_cdk_lib_aws_backup_TagCondition(tagCondition);
        this.resource = resource;
        this.tagCondition = tagCondition;
        this.construct = construct;
    }
    /**
     * Adds all supported resources in a construct.
     *
     * @param construct The construct containing resources to backup.
     * @stability stable
     */
    static fromConstruct(construct) {
        return new BackupResource(undefined, undefined, construct);
    }
    /**
     * A DynamoDB table.
     *
     * @stability stable
     */
    static fromDynamoDbTable(table) {
        jsiiDeprecationWarnings.aws_cdk_lib_aws_dynamodb_ITable(table);
        return BackupResource.fromArn(table.tableArn);
    }
    /**
     * An EC2 instance.
     *
     * @stability stable
     */
    static fromEc2Instance(instance) {
        jsiiDeprecationWarnings.aws_cdk_lib_aws_ec2_IInstance(instance);
        return BackupResource.fromArn(core_1.Stack.of(instance).formatArn({
            service: 'ec2',
            resource: 'instance',
            resourceName: instance.instanceId,
        }));
    }
    /**
     * An EFS file system.
     *
     * @stability stable
     */
    static fromEfsFileSystem(fileSystem) {
        jsiiDeprecationWarnings.aws_cdk_lib_aws_efs_IFileSystem(fileSystem);
        return BackupResource.fromArn(core_1.Stack.of(fileSystem).formatArn({
            service: 'elasticfilesystem',
            resource: 'file-system',
            resourceName: fileSystem.fileSystemId,
        }));
    }
    /**
     * A RDS database instance.
     *
     * @stability stable
     */
    static fromRdsDatabaseInstance(instance) {
        jsiiDeprecationWarnings.aws_cdk_lib_aws_rds_IDatabaseInstance(instance);
        return BackupResource.fromArn(instance.instanceArn);
    }
    /**
     * A list of ARNs or match patterns such as `arn:aws:ec2:us-east-1:123456789012:volume/*`.
     *
     * @stability stable
     */
    static fromArn(arn) {
        return new BackupResource(arn);
    }
    /**
     * A tag condition.
     *
     * @stability stable
     */
    static fromTag(key, value, operation) {
        jsiiDeprecationWarnings.aws_cdk_lib_aws_backup_TagOperation(operation);
        return new BackupResource(undefined, {
            key,
            value,
            operation,
        });
    }
}
exports.BackupResource = BackupResource;
_a = JSII_RTTI_SYMBOL_1;
BackupResource[_a] = { fqn: "aws-cdk-lib.aws_backup.BackupResource", version: "2.0.0" };
