// aws-cdk-lib/aws-lambda-destinations/lib/event-bridge.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.EventBridgeDestination = void 0;
const jsiiDeprecationWarnings = require("../../.warnings.jsii.js");
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const events = require("../../aws-events");
const core_1 = require("../../core");
/**
 * Use an Event Bridge event bus as a Lambda destination.
 *
 * If no event bus is specified, the default event bus is used.
 *
 * @stability stable
 */
class EventBridgeDestination {
    /**
     * @default - use the default event bus
     * @stability stable
     */
    constructor(eventBus) {
        this.eventBus = eventBus;
        jsiiDeprecationWarnings.aws_cdk_lib_aws_events_IEventBus(eventBus);
    }
    /**
     * Returns a destination configuration.
     *
     * @stability stable
     */
    bind(_scope, fn, _options) {
        jsiiDeprecationWarnings.aws_cdk_lib_aws_lambda_IFunction(fn);
        jsiiDeprecationWarnings.aws_cdk_lib_aws_lambda_DestinationOptions(_options);
        if (this.eventBus) {
            this.eventBus.grantPutEventsTo(fn);
            return {
                destination: this.eventBus.eventBusArn,
            };
        }
        const existingDefaultEventBus = _scope.node.tryFindChild('DefaultEventBus');
        let eventBus = existingDefaultEventBus || events.EventBus.fromEventBusArn(_scope, 'DefaultEventBus', core_1.Stack.of(fn).formatArn({
            service: 'events',
            resource: 'event-bus',
            resourceName: 'default',
        }));
        eventBus.grantPutEventsTo(fn);
        return {
            destination: eventBus.eventBusArn,
        };
    }
}
exports.EventBridgeDestination = EventBridgeDestination;
_a = JSII_RTTI_SYMBOL_1;
EventBridgeDestination[_a] = { fqn: "aws-cdk-lib.aws_lambda_destinations.EventBridgeDestination", version: "2.0.0" };
