// aws-cdk-lib/aws-fis/lib/fis.generated.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.CfnExperimentTemplate = void 0;
const jsiiDeprecationWarnings = require("../../.warnings.jsii.js");
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdk = require("../../core");
const cfn_parse = require("../../core/lib/cfn-parse");
/**
 * Determine whether the given properties match those of a `CfnExperimentTemplateProps`
 *
 * @param properties - the TypeScript properties of a `CfnExperimentTemplateProps`
 *
 * @returns the result of the validation.
 */
function CfnExperimentTemplatePropsValidator(properties) {
    if (!cdk.canInspect(properties)) {
        return cdk.VALIDATION_SUCCESS;
    }
    const errors = new cdk.ValidationResults();
    if (typeof properties !== 'object') {
        errors.collect(new cdk.ValidationResult('Expected an object, but received: ' + JSON.stringify(properties)));
    }
    errors.collect(cdk.propertyValidator('actions', cdk.hashValidator(CfnExperimentTemplate_ExperimentTemplateActionPropertyValidator))(properties.actions));
    errors.collect(cdk.propertyValidator('description', cdk.requiredValidator)(properties.description));
    errors.collect(cdk.propertyValidator('description', cdk.validateString)(properties.description));
    errors.collect(cdk.propertyValidator('roleArn', cdk.requiredValidator)(properties.roleArn));
    errors.collect(cdk.propertyValidator('roleArn', cdk.validateString)(properties.roleArn));
    errors.collect(cdk.propertyValidator('stopConditions', cdk.requiredValidator)(properties.stopConditions));
    errors.collect(cdk.propertyValidator('stopConditions', cdk.listValidator(CfnExperimentTemplate_ExperimentTemplateStopConditionPropertyValidator))(properties.stopConditions));
    errors.collect(cdk.propertyValidator('tags', cdk.requiredValidator)(properties.tags));
    errors.collect(cdk.propertyValidator('tags', cdk.hashValidator(cdk.validateString))(properties.tags));
    errors.collect(cdk.propertyValidator('targets', cdk.requiredValidator)(properties.targets));
    errors.collect(cdk.propertyValidator('targets', cdk.hashValidator(CfnExperimentTemplate_ExperimentTemplateTargetPropertyValidator))(properties.targets));
    return errors.wrap('supplied properties not correct for "CfnExperimentTemplateProps"');
}
/**
 * Renders the AWS CloudFormation properties of an `AWS::FIS::ExperimentTemplate` resource
 *
 * @param properties - the TypeScript properties of a `CfnExperimentTemplateProps`
 *
 * @returns the AWS CloudFormation properties of an `AWS::FIS::ExperimentTemplate` resource.
 */
// @ts-ignore TS6133
function cfnExperimentTemplatePropsToCloudFormation(properties) {
    if (!cdk.canInspect(properties)) {
        return properties;
    }
    CfnExperimentTemplatePropsValidator(properties).assertSuccess();
    return {
        Description: cdk.stringToCloudFormation(properties.description),
        RoleArn: cdk.stringToCloudFormation(properties.roleArn),
        StopConditions: cdk.listMapper(cfnExperimentTemplateExperimentTemplateStopConditionPropertyToCloudFormation)(properties.stopConditions),
        Tags: cdk.hashMapper(cdk.stringToCloudFormation)(properties.tags),
        Targets: cdk.hashMapper(cfnExperimentTemplateExperimentTemplateTargetPropertyToCloudFormation)(properties.targets),
        Actions: cdk.hashMapper(cfnExperimentTemplateExperimentTemplateActionPropertyToCloudFormation)(properties.actions),
    };
}
// @ts-ignore TS6133
function CfnExperimentTemplatePropsFromCloudFormation(properties) {
    properties = properties == null ? {} : properties;
    if (typeof properties !== 'object') {
        return new cfn_parse.FromCloudFormationResult(properties);
    }
    const ret = new cfn_parse.FromCloudFormationPropertyObject();
    ret.addPropertyResult('description', 'Description', cfn_parse.FromCloudFormation.getString(properties.Description));
    ret.addPropertyResult('roleArn', 'RoleArn', cfn_parse.FromCloudFormation.getString(properties.RoleArn));
    ret.addPropertyResult('stopConditions', 'StopConditions', cfn_parse.FromCloudFormation.getArray(CfnExperimentTemplateExperimentTemplateStopConditionPropertyFromCloudFormation)(properties.StopConditions));
    ret.addPropertyResult('tags', 'Tags', cfn_parse.FromCloudFormation.getMap(cfn_parse.FromCloudFormation.getString)(properties.Tags));
    ret.addPropertyResult('targets', 'Targets', cfn_parse.FromCloudFormation.getMap(CfnExperimentTemplateExperimentTemplateTargetPropertyFromCloudFormation)(properties.Targets));
    ret.addPropertyResult('actions', 'Actions', properties.Actions != null ? cfn_parse.FromCloudFormation.getMap(CfnExperimentTemplateExperimentTemplateActionPropertyFromCloudFormation)(properties.Actions) : undefined);
    ret.addUnrecognizedPropertiesAsExtra(properties);
    return ret;
}
/**
 * A CloudFormation `AWS::FIS::ExperimentTemplate`.
 *
 * @external
 * @cloudformationResource AWS::FIS::ExperimentTemplate
 * @link http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-fis-experimenttemplate.html
 */
class CfnExperimentTemplate extends cdk.CfnResource {
    /**
     * Create a new `AWS::FIS::ExperimentTemplate`.
     *
     * @param scope - scope in which this resource is defined.
     * @param id - scoped id of the resource.
     * @param props - resource properties.
     * @external
     */
    constructor(scope, id, props) {
        super(scope, id, { type: CfnExperimentTemplate.CFN_RESOURCE_TYPE_NAME, properties: props });
        jsiiDeprecationWarnings.aws_cdk_lib_aws_fis_CfnExperimentTemplateProps(props);
        cdk.requireProperty(props, 'description', this);
        cdk.requireProperty(props, 'roleArn', this);
        cdk.requireProperty(props, 'stopConditions', this);
        cdk.requireProperty(props, 'tags', this);
        cdk.requireProperty(props, 'targets', this);
        this.attrId = cdk.Token.asString(this.getAtt('Id'));
        this.description = props.description;
        this.roleArn = props.roleArn;
        this.stopConditions = props.stopConditions;
        this.tags = new cdk.TagManager(cdk.TagType.MAP, "AWS::FIS::ExperimentTemplate", props.tags, { tagPropertyName: 'tags' });
        this.targets = props.targets;
        this.actions = props.actions;
    }
    /**
     * A factory method that creates a new instance of this class from an object
     * containing the CloudFormation properties of this resource.
     * Used in the @aws-cdk/cloudformation-include module.
     *
     * @internal
     */
    static _fromCloudFormation(scope, id, resourceAttributes, options) {
        resourceAttributes = resourceAttributes || {};
        const resourceProperties = options.parser.parseValue(resourceAttributes.Properties);
        const propsResult = CfnExperimentTemplatePropsFromCloudFormation(resourceProperties);
        const ret = new CfnExperimentTemplate(scope, id, propsResult.value);
        for (const [propKey, propVal] of Object.entries(propsResult.extraProperties)) {
            ret.addPropertyOverride(propKey, propVal);
        }
        options.parser.handleAttributes(ret, resourceAttributes, id);
        return ret;
    }
    /**
     * Examines the CloudFormation resource and discloses attributes.
     *
     * @param inspector - tree inspector to collect and process attributes.
     * @external
     */
    inspect(inspector) {
        jsiiDeprecationWarnings.aws_cdk_lib_TreeInspector(inspector);
        inspector.addAttribute("aws:cdk:cloudformation:type", CfnExperimentTemplate.CFN_RESOURCE_TYPE_NAME);
        inspector.addAttribute("aws:cdk:cloudformation:props", this.cfnProperties);
    }
    /**
     * @external
     */
    get cfnProperties() {
        return {
            description: this.description,
            roleArn: this.roleArn,
            stopConditions: this.stopConditions,
            tags: this.tags.renderTags(),
            targets: this.targets,
            actions: this.actions,
        };
    }
    /**
     * @external
     */
    renderProperties(props) {
        return cfnExperimentTemplatePropsToCloudFormation(props);
    }
}
exports.CfnExperimentTemplate = CfnExperimentTemplate;
_a = JSII_RTTI_SYMBOL_1;
CfnExperimentTemplate[_a] = { fqn: "aws-cdk-lib.aws_fis.CfnExperimentTemplate", version: "2.0.0" };
/**
 * The CloudFormation resource type name for this resource class.
 *
 * @external
 */
CfnExperimentTemplate.CFN_RESOURCE_TYPE_NAME = "AWS::FIS::ExperimentTemplate";
/**
 * Determine whether the given properties match those of a `ExperimentTemplateActionProperty`
 *
 * @param properties - the TypeScript properties of a `ExperimentTemplateActionProperty`
 *
 * @returns the result of the validation.
 */
function CfnExperimentTemplate_ExperimentTemplateActionPropertyValidator(properties) {
    if (!cdk.canInspect(properties)) {
        return cdk.VALIDATION_SUCCESS;
    }
    const errors = new cdk.ValidationResults();
    if (typeof properties !== 'object') {
        errors.collect(new cdk.ValidationResult('Expected an object, but received: ' + JSON.stringify(properties)));
    }
    errors.collect(cdk.propertyValidator('actionId', cdk.requiredValidator)(properties.actionId));
    errors.collect(cdk.propertyValidator('actionId', cdk.validateString)(properties.actionId));
    errors.collect(cdk.propertyValidator('description', cdk.validateString)(properties.description));
    errors.collect(cdk.propertyValidator('parameters', cdk.hashValidator(cdk.validateString))(properties.parameters));
    errors.collect(cdk.propertyValidator('startAfter', cdk.listValidator(cdk.validateString))(properties.startAfter));
    errors.collect(cdk.propertyValidator('targets', cdk.hashValidator(cdk.validateString))(properties.targets));
    return errors.wrap('supplied properties not correct for "ExperimentTemplateActionProperty"');
}
/**
 * Renders the AWS CloudFormation properties of an `AWS::FIS::ExperimentTemplate.ExperimentTemplateAction` resource
 *
 * @param properties - the TypeScript properties of a `ExperimentTemplateActionProperty`
 *
 * @returns the AWS CloudFormation properties of an `AWS::FIS::ExperimentTemplate.ExperimentTemplateAction` resource.
 */
// @ts-ignore TS6133
function cfnExperimentTemplateExperimentTemplateActionPropertyToCloudFormation(properties) {
    if (!cdk.canInspect(properties)) {
        return properties;
    }
    CfnExperimentTemplate_ExperimentTemplateActionPropertyValidator(properties).assertSuccess();
    return {
        ActionId: cdk.stringToCloudFormation(properties.actionId),
        Description: cdk.stringToCloudFormation(properties.description),
        Parameters: cdk.hashMapper(cdk.stringToCloudFormation)(properties.parameters),
        StartAfter: cdk.listMapper(cdk.stringToCloudFormation)(properties.startAfter),
        Targets: cdk.hashMapper(cdk.stringToCloudFormation)(properties.targets),
    };
}
// @ts-ignore TS6133
function CfnExperimentTemplateExperimentTemplateActionPropertyFromCloudFormation(properties) {
    if (cdk.isResolvableObject(properties)) {
        return new cfn_parse.FromCloudFormationResult(properties);
    }
    properties = properties == null ? {} : properties;
    if (typeof properties !== 'object') {
        return new cfn_parse.FromCloudFormationResult(properties);
    }
    const ret = new cfn_parse.FromCloudFormationPropertyObject();
    ret.addPropertyResult('actionId', 'ActionId', cfn_parse.FromCloudFormation.getString(properties.ActionId));
    ret.addPropertyResult('description', 'Description', properties.Description != null ? cfn_parse.FromCloudFormation.getString(properties.Description) : undefined);
    ret.addPropertyResult('parameters', 'Parameters', properties.Parameters != null ? cfn_parse.FromCloudFormation.getMap(cfn_parse.FromCloudFormation.getString)(properties.Parameters) : undefined);
    ret.addPropertyResult('startAfter', 'StartAfter', properties.StartAfter != null ? cfn_parse.FromCloudFormation.getStringArray(properties.StartAfter) : undefined);
    ret.addPropertyResult('targets', 'Targets', properties.Targets != null ? cfn_parse.FromCloudFormation.getMap(cfn_parse.FromCloudFormation.getString)(properties.Targets) : undefined);
    ret.addUnrecognizedPropertiesAsExtra(properties);
    return ret;
}
/**
 * Determine whether the given properties match those of a `ExperimentTemplateStopConditionProperty`
 *
 * @param properties - the TypeScript properties of a `ExperimentTemplateStopConditionProperty`
 *
 * @returns the result of the validation.
 */
function CfnExperimentTemplate_ExperimentTemplateStopConditionPropertyValidator(properties) {
    if (!cdk.canInspect(properties)) {
        return cdk.VALIDATION_SUCCESS;
    }
    const errors = new cdk.ValidationResults();
    if (typeof properties !== 'object') {
        errors.collect(new cdk.ValidationResult('Expected an object, but received: ' + JSON.stringify(properties)));
    }
    errors.collect(cdk.propertyValidator('source', cdk.requiredValidator)(properties.source));
    errors.collect(cdk.propertyValidator('source', cdk.validateString)(properties.source));
    errors.collect(cdk.propertyValidator('value', cdk.validateString)(properties.value));
    return errors.wrap('supplied properties not correct for "ExperimentTemplateStopConditionProperty"');
}
/**
 * Renders the AWS CloudFormation properties of an `AWS::FIS::ExperimentTemplate.ExperimentTemplateStopCondition` resource
 *
 * @param properties - the TypeScript properties of a `ExperimentTemplateStopConditionProperty`
 *
 * @returns the AWS CloudFormation properties of an `AWS::FIS::ExperimentTemplate.ExperimentTemplateStopCondition` resource.
 */
// @ts-ignore TS6133
function cfnExperimentTemplateExperimentTemplateStopConditionPropertyToCloudFormation(properties) {
    if (!cdk.canInspect(properties)) {
        return properties;
    }
    CfnExperimentTemplate_ExperimentTemplateStopConditionPropertyValidator(properties).assertSuccess();
    return {
        Source: cdk.stringToCloudFormation(properties.source),
        Value: cdk.stringToCloudFormation(properties.value),
    };
}
// @ts-ignore TS6133
function CfnExperimentTemplateExperimentTemplateStopConditionPropertyFromCloudFormation(properties) {
    if (cdk.isResolvableObject(properties)) {
        return new cfn_parse.FromCloudFormationResult(properties);
    }
    properties = properties == null ? {} : properties;
    if (typeof properties !== 'object') {
        return new cfn_parse.FromCloudFormationResult(properties);
    }
    const ret = new cfn_parse.FromCloudFormationPropertyObject();
    ret.addPropertyResult('source', 'Source', cfn_parse.FromCloudFormation.getString(properties.Source));
    ret.addPropertyResult('value', 'Value', properties.Value != null ? cfn_parse.FromCloudFormation.getString(properties.Value) : undefined);
    ret.addUnrecognizedPropertiesAsExtra(properties);
    return ret;
}
/**
 * Determine whether the given properties match those of a `ExperimentTemplateTargetProperty`
 *
 * @param properties - the TypeScript properties of a `ExperimentTemplateTargetProperty`
 *
 * @returns the result of the validation.
 */
function CfnExperimentTemplate_ExperimentTemplateTargetPropertyValidator(properties) {
    if (!cdk.canInspect(properties)) {
        return cdk.VALIDATION_SUCCESS;
    }
    const errors = new cdk.ValidationResults();
    if (typeof properties !== 'object') {
        errors.collect(new cdk.ValidationResult('Expected an object, but received: ' + JSON.stringify(properties)));
    }
    errors.collect(cdk.propertyValidator('filters', cdk.listValidator(CfnExperimentTemplate_ExperimentTemplateTargetFilterPropertyValidator))(properties.filters));
    errors.collect(cdk.propertyValidator('resourceArns', cdk.listValidator(cdk.validateString))(properties.resourceArns));
    errors.collect(cdk.propertyValidator('resourceTags', cdk.hashValidator(cdk.validateString))(properties.resourceTags));
    errors.collect(cdk.propertyValidator('resourceType', cdk.requiredValidator)(properties.resourceType));
    errors.collect(cdk.propertyValidator('resourceType', cdk.validateString)(properties.resourceType));
    errors.collect(cdk.propertyValidator('selectionMode', cdk.requiredValidator)(properties.selectionMode));
    errors.collect(cdk.propertyValidator('selectionMode', cdk.validateString)(properties.selectionMode));
    return errors.wrap('supplied properties not correct for "ExperimentTemplateTargetProperty"');
}
/**
 * Renders the AWS CloudFormation properties of an `AWS::FIS::ExperimentTemplate.ExperimentTemplateTarget` resource
 *
 * @param properties - the TypeScript properties of a `ExperimentTemplateTargetProperty`
 *
 * @returns the AWS CloudFormation properties of an `AWS::FIS::ExperimentTemplate.ExperimentTemplateTarget` resource.
 */
// @ts-ignore TS6133
function cfnExperimentTemplateExperimentTemplateTargetPropertyToCloudFormation(properties) {
    if (!cdk.canInspect(properties)) {
        return properties;
    }
    CfnExperimentTemplate_ExperimentTemplateTargetPropertyValidator(properties).assertSuccess();
    return {
        Filters: cdk.listMapper(cfnExperimentTemplateExperimentTemplateTargetFilterPropertyToCloudFormation)(properties.filters),
        ResourceArns: cdk.listMapper(cdk.stringToCloudFormation)(properties.resourceArns),
        ResourceTags: cdk.hashMapper(cdk.stringToCloudFormation)(properties.resourceTags),
        ResourceType: cdk.stringToCloudFormation(properties.resourceType),
        SelectionMode: cdk.stringToCloudFormation(properties.selectionMode),
    };
}
// @ts-ignore TS6133
function CfnExperimentTemplateExperimentTemplateTargetPropertyFromCloudFormation(properties) {
    if (cdk.isResolvableObject(properties)) {
        return new cfn_parse.FromCloudFormationResult(properties);
    }
    properties = properties == null ? {} : properties;
    if (typeof properties !== 'object') {
        return new cfn_parse.FromCloudFormationResult(properties);
    }
    const ret = new cfn_parse.FromCloudFormationPropertyObject();
    ret.addPropertyResult('filters', 'Filters', properties.Filters != null ? cfn_parse.FromCloudFormation.getArray(CfnExperimentTemplateExperimentTemplateTargetFilterPropertyFromCloudFormation)(properties.Filters) : undefined);
    ret.addPropertyResult('resourceArns', 'ResourceArns', properties.ResourceArns != null ? cfn_parse.FromCloudFormation.getStringArray(properties.ResourceArns) : undefined);
    ret.addPropertyResult('resourceTags', 'ResourceTags', properties.ResourceTags != null ? cfn_parse.FromCloudFormation.getMap(cfn_parse.FromCloudFormation.getString)(properties.ResourceTags) : undefined);
    ret.addPropertyResult('resourceType', 'ResourceType', cfn_parse.FromCloudFormation.getString(properties.ResourceType));
    ret.addPropertyResult('selectionMode', 'SelectionMode', cfn_parse.FromCloudFormation.getString(properties.SelectionMode));
    ret.addUnrecognizedPropertiesAsExtra(properties);
    return ret;
}
/**
 * Determine whether the given properties match those of a `ExperimentTemplateTargetFilterProperty`
 *
 * @param properties - the TypeScript properties of a `ExperimentTemplateTargetFilterProperty`
 *
 * @returns the result of the validation.
 */
function CfnExperimentTemplate_ExperimentTemplateTargetFilterPropertyValidator(properties) {
    if (!cdk.canInspect(properties)) {
        return cdk.VALIDATION_SUCCESS;
    }
    const errors = new cdk.ValidationResults();
    if (typeof properties !== 'object') {
        errors.collect(new cdk.ValidationResult('Expected an object, but received: ' + JSON.stringify(properties)));
    }
    errors.collect(cdk.propertyValidator('path', cdk.requiredValidator)(properties.path));
    errors.collect(cdk.propertyValidator('path', cdk.validateString)(properties.path));
    errors.collect(cdk.propertyValidator('values', cdk.requiredValidator)(properties.values));
    errors.collect(cdk.propertyValidator('values', cdk.listValidator(cdk.validateString))(properties.values));
    return errors.wrap('supplied properties not correct for "ExperimentTemplateTargetFilterProperty"');
}
/**
 * Renders the AWS CloudFormation properties of an `AWS::FIS::ExperimentTemplate.ExperimentTemplateTargetFilter` resource
 *
 * @param properties - the TypeScript properties of a `ExperimentTemplateTargetFilterProperty`
 *
 * @returns the AWS CloudFormation properties of an `AWS::FIS::ExperimentTemplate.ExperimentTemplateTargetFilter` resource.
 */
// @ts-ignore TS6133
function cfnExperimentTemplateExperimentTemplateTargetFilterPropertyToCloudFormation(properties) {
    if (!cdk.canInspect(properties)) {
        return properties;
    }
    CfnExperimentTemplate_ExperimentTemplateTargetFilterPropertyValidator(properties).assertSuccess();
    return {
        Path: cdk.stringToCloudFormation(properties.path),
        Values: cdk.listMapper(cdk.stringToCloudFormation)(properties.values),
    };
}
// @ts-ignore TS6133
function CfnExperimentTemplateExperimentTemplateTargetFilterPropertyFromCloudFormation(properties) {
    if (cdk.isResolvableObject(properties)) {
        return new cfn_parse.FromCloudFormationResult(properties);
    }
    properties = properties == null ? {} : properties;
    if (typeof properties !== 'object') {
        return new cfn_parse.FromCloudFormationResult(properties);
    }
    const ret = new cfn_parse.FromCloudFormationPropertyObject();
    ret.addPropertyResult('path', 'Path', cfn_parse.FromCloudFormation.getString(properties.Path));
    ret.addPropertyResult('values', 'Values', cfn_parse.FromCloudFormation.getStringArray(properties.Values));
    ret.addUnrecognizedPropertiesAsExtra(properties);
    return ret;
}
