// aws-cdk-lib/aws-globalaccelerator-endpoints/lib/eip.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.CfnEipEndpoint = void 0;
const jsiiDeprecationWarnings = require("../../.warnings.jsii.js");
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const core_1 = require("../../core");
const _util_1 = require("./_util");
/**
 * Use an EC2 Instance as a Global Accelerator Endpoint.
 *
 * @stability stable
 */
class CfnEipEndpoint {
    /**
     * @stability stable
     */
    constructor(eip, options = {}) {
        this.eip = eip;
        this.options = options;
        jsiiDeprecationWarnings.aws_cdk_lib_aws_ec2_CfnEIP(eip);
        jsiiDeprecationWarnings.aws_cdk_lib_aws_globalaccelerator_endpoints_CfnEipEndpointProps(options);
        _util_1.validateWeight(options.weight);
        this.region = core_1.Stack.of(eip).region;
    }
    /**
     * Render the endpoint to an endpoint configuration.
     *
     * @stability stable
     */
    renderEndpointConfiguration() {
        return {
            endpointId: this.eip.attrAllocationId,
            weight: this.options.weight,
        };
    }
}
exports.CfnEipEndpoint = CfnEipEndpoint;
_a = JSII_RTTI_SYMBOL_1;
CfnEipEndpoint[_a] = { fqn: "aws-cdk-lib.aws_globalaccelerator_endpoints.CfnEipEndpoint", version: "2.0.0" };
