// aws-cdk-lib/aws-globalaccelerator-endpoints/lib/instance.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.InstanceEndpoint = void 0;
const jsiiDeprecationWarnings = require("../../.warnings.jsii.js");
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const _util_1 = require("./_util");
/**
 * Use an EC2 Instance as a Global Accelerator Endpoint.
 *
 * @stability stable
 */
class InstanceEndpoint {
    /**
     * @stability stable
     */
    constructor(instance, options = {}) {
        this.instance = instance;
        this.options = options;
        jsiiDeprecationWarnings.aws_cdk_lib_aws_ec2_IInstance(instance);
        jsiiDeprecationWarnings.aws_cdk_lib_aws_globalaccelerator_endpoints_InstanceEndpointProps(options);
        _util_1.validateWeight(options.weight);
        this.region = instance.env.region;
    }
    /**
     * Render the endpoint to an endpoint configuration.
     *
     * @stability stable
     */
    renderEndpointConfiguration() {
        return {
            endpointId: this.instance.instanceId,
            weight: this.options.weight,
            clientIpPreservationEnabled: this.options.preserveClientIp,
        };
    }
}
exports.InstanceEndpoint = InstanceEndpoint;
_a = JSII_RTTI_SYMBOL_1;
InstanceEndpoint[_a] = { fqn: "aws-cdk-lib.aws_globalaccelerator_endpoints.InstanceEndpoint", version: "2.0.0" };
